Main.hs

```diff
@@ -4,26 +4,85 @@ import Control.Monad (unless)
 import Text.Printf (printf)
 
 data Expr
+  = Number   Double
+  | Sqrt     Expr
+  | Add      Expr Expr
+  | Subtract Expr Expr
+  | Multiply Expr Expr
+  | Divide   Expr Expr
+  | Power    Expr Expr
 
 instance Show Expr where
-  show = undefined
+  show (Number x)     = show x
+  show (Sqrt x)       = "sqrt(" ++ show x ++ ")"
+  show (Add x y)      = "(" ++ show x ++ " + " ++ show y ++ ")"
+  show (Subtract x y) = "(" ++ show x ++ " - " ++ show y ++ ")"
+  show (Multiply x y) = "(" ++ show x ++ " * " ++ show y ++ ")"
+  show (Divide x y)   = "(" ++ show x ++ " / " ++ show y ++ ")"
+  show (Power x y)    = "(" ++ show x ++ " ^ " ++ show y ++ ")"
 
 instance Eq Expr where
-  (==) = undefined
+  (Number x)       == (Number y)       = x == y
+  (Sqrt x)         == (Sqrt y)         = x == y
+  (Add x1 y1)      == (Add x2 y2)      = x1 == x2 && y1 == y2
+  (Subtract x1 y1) == (Subtract x2 y2) = x1 == x2 && y1 == y2
+  (Multiply x1 y1) == (Multiply x2 y2) = x1 == x2 && y1 == y2
+  (Divide x1 y1)   == (Divide x2 y2)   = x1 == x2 && y1 == y2
+  (Power x1 y1)    == (Power x2 y2)    = x1 == x2 && y1 == y2
+  _ == _ = False
```

Comment:
Whenever your implementation of a derivable type class is the same as the default one, just use deriving. 
Otherwise, the reader of your code would waste time trying to figure out what the difference is.